### DISCUSSION TITLE
Compilation failed for Citext PostgreSQL type #3727 - follow-up

### DISCUSSION THREAD

**Discussion by nerdunit** [diesel-rs/diesel#3728] [Q&A]
https://github.com/diesel-rs/diesel/issues/3727

I did do the implementation [Mingun](https://github.com/Mingun) and [weiznich](https://github.com/weiznich) suggested
```rs
    impl FromSql<Citext, Pg> for String {
        fn from_sql(bytes: diesel::backend::RawValue<Pg>) -> deserialize::Result<Self> {
            Ok(String::from_utf8_lossy(bytes.as_bytes()).to_string())
        }

        fn from_nullable_sql(bytes: Option<diesel::backend::RawValue<Pg>>) -> deserialize::Result<Self> {
            match bytes {
                Some(bytes) => Self::from_sql(bytes),
                None => Ok(String::default()),
            }
        }
    }

    impl ToSql<Citext, Pg> for String {
        fn to_sql<'a>(&'a self, out: &mut Output<'a, '_, Pg>) -> serialize::Result {
            out.write_all(self.as_bytes())?;
            Ok(IsNull::No)
        }
    }
```
However this still doesn't compile...
```
   |
41 |     username: Citext,
   |     ^^^^^^^^ the trait `diesel::Expression` is not implemented for `Citext`
   |
```

**Comment by Mingun** [comment]
Because you should use `username: String`. `Citext` for `table!`

**Reply by nerdunit** [reply]
THank you, however I get ``error[E0277]: the trait bound `std::string::String: diesel::Expression` is not satisfied`` using String

**Reply by nerdunit** [reply]
@Mingun I kept trying different things, but it doesn't work at all, neither with username: String neither with username: Citext

**Reply by Mingun** [reply]
I think you need to show all relevant code, and where exactly you see the error (maybe a link to the project if it is public). I personally does not use diesel, only watching some interesting projects.

**Reply by PaulDotSH** [reply]
Thanks... I temporarely commented that struct and my code was wrong anyway, I found this crate https://github.com/gregelenbaas/diesel-citext however it is outdated and I'm not sure how to fix it... I would use sqlx or another orm but i need diesel's performance

**Comment by weiznich** [comment]
Based on the provided information it's impossible to really help you. Try to provide at least the following things:

* All the relevant code to reproduce your problem
* The complete error message as emitted by rustc and not only a part of the error message.

As general rule for asking questions: If you expect someone to spend time on debugging and answering your questions you should spend time to ask well written questions that make answering them easy. I personally try to spend a similar effort on answering question than the OP spend on writing them. So just dumping a piece of code and an incomplete error message is a good way of not getting a long answer..

Otherwise I have the strong supposition that your implementation is just incomplete. Try following the documentation as shown in [this](https://docs.diesel.rs/2.1.x/diesel/deserialize/trait.FromSql.html#examples) and [this](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html#examples) example. Note the diesel specific derives on the rust side types there!
